src/number_theory/fermat4.lean

```diff
@@ -0,0 +1,335 @@
+end
+
+lemma int.coprime_of_sqr_sum' {m r s : ℤ} (h : m = r ^ 2 + s ^ 2) (h2 : int.gcd r s = 1) :
+  int.gcd m (r * s) = 1 :=
+begin
+  dunfold int.gcd,
+  rw int.nat_abs_mul,
+  apply nat.coprime.mul_right (int.coprime_of_sqr_sum h h2),
+  rw add_comm at h,
+  rw [int.gcd_comm r s] at h2,
+  exact int.coprime_of_sqr_sum h h2
+end
+
+namespace fermat_42
+
+-- If we have a solution to a ^ 4 + b ^ 4 = c ^ 2, we can construct a smaller one. This
+-- implies there can't be a smallest solution.
+lemma not_minimal {a b c : ℤ}
+  (h : minimal a b c) (ha2 : a % 2 = 1) (hc : 0 < c) : false :=
+begin
+  -- Use the fact that a ^ 2, b ^ 2, c form a pythagorean triple to obtain m and n such that
+  -- a ^ 2 = m ^ 2 - n ^ 2, b ^ 2 = 2 * m * n and c = m ^ 2 + n ^ 2
+  -- first the formula:
+  have ht : pythagorean_triple (a ^ 2) (b ^ 2) c,
+  { calc ((a ^ 2) * (a ^ 2)) + ((b ^ 2) * (b ^ 2))
+        = a ^ 4 + b ^ 4 : by ring
+    ... = c ^ 2 : by rw h.1.2.2
+    ... = c * c : by rw pow_two
+  },
+  -- coprime requirement:
+  have h2 : int.gcd (a ^ 2) (b ^ 2) = 1,
+  { change nat.coprime (int.nat_abs (a ^ 2)) (int.nat_abs (b ^ 2)),
+    rw [pow_two, int.nat_abs_mul],
+    apply nat.coprime.mul;
+    { rw [pow_two, int.nat_abs_mul], apply nat.coprime.mul_right;
+      apply (coprime_of_minimal h) } },
+  -- in order to reduce the possibilities we get from the classification of pythagorean triples
+  -- it helps if we know the parity of a ^ 2 (and the sign of c):
+  have ha22 : a ^ 2 % 2 = 1, { rw [pow_two, int.mul_mod, ha2], norm_num },
+  obtain ⟨m, n, ht1, ht2, ht3, ht4, ht5, ht6⟩ :=
+    pythagorean_triple.coprime_classification' ht h2 ha22 hc,
+  -- Now a, n, m form a pythagorean triple and so we can obtain r and s such that
+  -- a = r ^ 2 - s ^ 2, n = 2 * r * s and m = r ^ 2 + s ^ 2
+  -- formula:
+  have htt : pythagorean_triple a n m,
+  { delta pythagorean_triple,
+    rw [← pow_two, ← pow_two, ← pow_two], exact (add_eq_of_eq_sub ht1) },
+  -- a and n are coprime, because a ^ 2 = m ^ 2 - n ^ 2 and m and n are coprime.
+  have h3 : int.gcd a n = 1,
+  { have ha : (int.gcd a n : ℤ) ^ 2 ∣ a ^ 2,
+    { exact (int.pow_dvd_pow_iff (dec_trivial : 0 < 2)).mpr (int.gcd_dvd_left a n),},
+    have hn : (int.gcd a n : ℤ) ^ 2 ∣ n ^ 2,
+    { exact (int.pow_dvd_pow_iff (dec_trivial : 0 < 2)).mpr (int.gcd_dvd_right a n),},
+    have hm : (int.gcd a n : ℤ) ^ 2 ∣ m ^ 2,
+    { rw [← (add_eq_of_eq_sub ht1)], exact dvd_add ha hn },
+    replace hm : (int.gcd a n : ℤ) ∣ m, { exact (int.pow_dvd_pow_iff (dec_trivial : 0 < 2)).mp hm},
+    apply nat.eq_one_of_dvd_one,
+    rw [←ht4], apply int.coe_nat_dvd.mp, exact int.dvd_gcd hm (int.gcd_dvd_right a n) },
+  -- m is positive because b is non-zero and b ^ 2 = 2 * m * n and we already have 0 ≤ m.
+  have hb20 : b ^ 2 ≠ 0, { exact mt pow_eq_zero h.1.2.1 },
+  have h4 : 0 < m,
+  { apply lt_of_le_of_ne ht6,
+    by_contradiction hm0,
+    revert hb20,
+    rw [ht2, ← (not_not.mp hm0)], simp },
```

Comment:
```suggestion
  { apply lt_of_le_of_ne ht6,
    rintro rfl,
    revert hb20,
    rw ht2, simp },
```

Reply:
You don't need to prove X != 0 by contradiction; the definition of `X \ne 0` is `X = 0 -> false`. And again the `rintro rfl` trick saves you a rewrite.